toolchain/check/check.cpp

```diff
@@ -231,14 +250,360 @@ static auto InitPackageScopeAndImports(Context& context, UnitInfo& unit_info)
+// need to reorder.
+struct NextInlineMethodCache {
+  const Parse::Tree* tree;
+  Parse::InlineMethodIndex index = Parse::InlineMethodIndex::Invalid;
+  Parse::NodeId start_id = Parse::NodeId::Invalid;
+
+  // Set the specified inline method index as being the next one to parse.
+  auto SkipTo(Parse::InlineMethodIndex next_index) -> void {
+    index = next_index;
+    if (static_cast<std::size_t>(index.index) ==
+        tree->inline_methods().size()) {
+      start_id = Parse::NodeId::Invalid;
+    } else {
+      start_id = tree->inline_methods().Get(index).start_id;
+    }
+  }
+};
+}  // namespace
+
+// Determines whether we are currently declaring a name in a scope in which
+// method definitions are delayed. When entering another delayed method scope,
+// the inner declaration's methods are parsed at the end of the outer
+// declaration, not the inner one.  For example:
+//
+// ```
+// class A {
+//   class B {
+//     fn F() -> A { return {}; }
+//   }
+// } // A.B.F is parsed here, with A complete.
+//
+// fn F() {
+//   class C {
+//     fn G() {}
+//   } // C.G is parsed here.
+// }
+// ```
+static auto IsInDelayedMethodScope(Context& context) -> bool {
+  auto inst_id = context.name_scopes().GetInstIdIfValid(
+      context.decl_name_stack().PeekTargetScope());
+  if (!inst_id.is_valid()) {
+    return false;
+  }
+  switch (context.insts().Get(inst_id).kind()) {
+    case SemIR::ClassDecl::Kind:
+    case SemIR::ImplDecl::Kind:
+    case SemIR::InterfaceDecl::Kind:
+      // TODO: Named constraints, mixins.
+      return true;
+
+    default:
+      return false;
+  }
+}
+
+// Determines whether this node kind is the start of a delayed method scope.
+static auto IsStartOfDelayedMethodScope(Parse::NodeKind kind) -> bool {
+  switch (kind) {
+    case Parse::NodeKind::ClassDefinitionStart:
+    case Parse::NodeKind::ImplDefinitionStart:
+    case Parse::NodeKind::InterfaceDefinitionStart:
+    case Parse::NodeKind::NamedConstraintDefinitionStart:
+      // TODO: Mixins.
+      return true;
+    default:
+      return false;
+  }
+}
+
+// Determines whether this node kind is the end of a delayed method scope.
+static auto IsEndOfDelayedMethodScope(Parse::NodeKind kind) -> bool {
+  switch (kind) {
+    case Parse::NodeKind::ClassDefinition:
+    case Parse::NodeKind::ImplDefinition:
+    case Parse::NodeKind::InterfaceDefinition:
+    case Parse::NodeKind::NamedConstraintDefinition:
+      // TODO: Mixins.
+      return true;
+    default:
+      return false;
+  }
+}
+
+namespace {
+// A worklist of pending tasks to perform to parse skipped inline method
+// bodies.
+struct InlineMethodWorklist {
+  // A worklist task that indicates we should parse a skipped method.
+  struct ParseSkippedMethod {
+    // The method that we skipped.
+    Parse::InlineMethodIndex method_index;
+    // The suspended function. Only present when `method_index` is valid.
+    std::optional<SuspendedFunction> sus_fn;
```

Comment:
Should be less sus now. I don't think we need an abbreviation here, really.

For other names: I think "delayed", "deferred", and "relocated" do a less good job of capturing "already started but not active right now" for me. I could go with "paused" (and `Pause` instead of `Suspend`) if that's better? Looking through a thesaurus, I also considered "interrupted", which I don't think works well because of other potential meanings, and "adjourned", which seems a bit too obscure.

Reply:
"Suspend" seems fine to me, was just trying to think if you wanted something with a clean abbreviation.

I don't know if "deferred" would work better for you now,j given the other changes. Maybe it's more complex; PartialDeferredFunction? :)

Reply:
I think we want two different words here. We're using "deferred" in a timeless sense to refer to functions and scopes that need special processing, whether or not that special processing has already begun or finished, whereas we're using "suspended" to refer to things that are suspended *right now*. For example, we have comments like "Suspend the current deferred definition scope" that would be confusing if we used the same word for both senses.